R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+#'     that the *name* of the function is passed!) the resulting m/z is
+#'     determined as an intensity-weighted mean of spectras' m/z values.
+#'
+#' @param intensityFun `function` to aggregate the intensity values per m/z
+#'     group.
+#'
+#' @param mzd `numeric(1)` defining the maximal m/z difference below which
+#'     values are grouped. If not provided, this value is estimated from the
+#'     distribution of differences of m/z values from the provided spectra
+#'     (see details).
+#' 
+#' @return
+#'
+#' `Spectrum` with m/z and intensity values representing the aggregated values
+#' across the provided spectra. The returned spectrum contains the same number
+#' of m/z and intensity pairs than the spectrum with index `main` in `x`, also
+#' all other related information is taken from this spectrum.
+#'
+#' @author Johannes Rainer
+#'
+#' @seealso
+#'
+#' [estimateMzScattering()] for a function to estimate m/z scattering
+#' in consecutive scans.
+#'
+#' [estimateMzResolution()] for a function estimating the m/z resolution of
+#' a spectrum.
+#' 
+#' [combineSpectraMovingWindow()] for the function to combine consecutive
+#' spectra of an `MSnExp` object using a moving window approach.
+#' 
+#' @md
+#'
+#' @examples
+#'
+#' library(MSnbase)
+#' ## Create 3 example profile-mode spectra with a resolution of 0.1 and small
+#' ## random variations on these m/z values on consecutive scans.
+#' set.seed(123)
+#' mzs <- seq(1, 20, 0.1)
+#' ints1 <- abs(rnorm(length(mzs), 10))
+#' ints1[11:20] <- c(15, 30, 90, 200, 500, 300, 100, 70, 40, 20) # add peak
+#' ints2 <- abs(rnorm(length(mzs), 10))
+#' ints2[11:20] <- c(15, 30, 60, 120, 300, 200, 90, 60, 30, 23)
+#' ints3 <- abs(rnorm(length(mzs), 10))
+#' ints3[11:20] <- c(13, 20, 50, 100, 200, 100, 80, 40, 30, 20)
+#'
+#' ## Create the spectra.
+#' sp1 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints1)
+#' sp2 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints2)
+#' sp3 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.009),
+#'     intensity = ints3)
+#'
+#' ## Combine the spectra
+#' sp_agg <- combineSpectra(list(sp1, sp2, sp3))
+#'
+#' ## Plot the spectra before and after combining
+#' par(mfrow = c(2, 1), mar = c(4.3, 4, 1, 1))
+#' plot(mz(sp1), intensity(sp1), xlim = range(mzs[5:25]), type = "h", col = "red")
+#' points(mz(sp2), intensity(sp2), type = "h", col = "green")
+#' points(mz(sp3), intensity(sp3), type = "h", col = "blue")
+#' plot(mz(sp_agg), intensity(sp_agg), xlim = range(mzs[5:25]), type = "h",
+#'     col = "black")
+combineSpectra <- function(x, mzFun = base::mean, intensityFun = base::mean,
+                           main = ceiling(median(1:length(x))),
+                           mzd) {
+    if (length(unique(unlist(lapply(x, function(z) z@msLevel)))) != 1)
+        stop("Can only combine spectra with the same MS level")
+    mzs <- lapply(x, function(z) z@mz)
```

Comment:
Yes, I'm using `@` here to avoid method dispatching. If you are calling this in a loop over e.g. 3000 spectra there is quite some performance improvement. It is something that I usually don't like to do (I'm against directly accessing slot values) - but sometimes, especially on large lists of S4 objects it can improve performance.

I'm using it now also in the `estimateMzScattering`.

Reply:
Ok, fine.